tests/fips/openssl/openssl_fips_alglist.pm

```diff
@@ -15,64 +15,83 @@ use base "consoletest";
-    else {
-        record_soft_failure('jsc#SLE-19640: openssl version is outdated and need to be updated over 1.1.1+ for SLE15-SP4');
-    }
+    my @openssl_output = split /\n/, script_output("$openssl_binary $list_command");
 
-    # Seperate the diffrent openssl command usage between SLE12 and SLE 15-SP{0..5} and SLE 15-SP6+
-    if (is_sle('<15')) {
-        # List message digest algorithms in fips mode
-        # only SHA1 and SHA2 (224, 256, 384, 512) are approved in fips mode
-        # Note: DSA is short of DSA-SHA1, so it is also valid item
-        validate_script_output
-"echo -n 'Invalid Hash: '; openssl list-message-digest-algorithms | sed -e '/SHA1/d' -e '/SHA224/d' -e '/SHA256/d' -e '/SHA384/d' -e '/SHA512/d' -e '/DSA/d' | wc -l",
-          sub { m/^Invalid Hash: 0$/ };
-
-        # List public key algorithms in fips mode
-        # only RSA, DSA, ECDSA, EC DH, CMAC and HMAC are approved in fips mode
-        validate_script_output
-"echo -n 'Invalid Pubkey: '; openssl list-public-key-algorithms | grep '^Name' | sed -e '/RSA/d' -e '/rsa/d' -e '/DSA/d' -e '/dsa/d' -e '/EC/d' -e '/DH/d' -e '/HMAC/d' -e '/CMAC/d' | wc -l",
-          sub { m/^Invalid Pubkey: 0$/ };
-    } elsif (is_sle('<15-SP6')) {
-        eval {
-            validate_script_output
-"echo -n 'Invalid Hash: '; openssl list -digest-algorithms | sed -e '/SHA1/d' -e '/SHA224/d' -e '/SHA256/d' -e '/SHA384/d' -e '/SHA512/d' -e '/DSA/d' -e '/SHA3-224/d' -e '/SHA3-256/d' -e '/SHA3-384/d' -e '/SHA3-512/d' -e '/SHAKE128/d' -e '/SHAKE256/d' | wc -l",
-              sub { m/^Invalid Hash: 0$/ };
-
-            validate_script_output
-"echo -n 'Invalid Pubkey: '; openssl list -public-key-algorithms | grep '^Name' | sed -e '/RSA/d' -e '/rsa/d' -e '/DSA/d' -e '/dsa/d' -e '/EC/d' -e '/DH/d' -e '/HMAC/d' -e '/CMAC/d' | wc -l",
-              sub { m/^Invalid Pubkey: 0$/ };
-        }
-    } else {
-        eval {
-            validate_script_output
-"echo -n 'Invalid Hash: '; openssl list -digest-algorithms | grep fips | sed -e '/SHA1/d' -e '/SHA224/d' -e '/SHA256/d' -e '/SHA384/d' -e '/SHA512/d' -e '/DSA/d' -e '/SHA3-224/d' -e '/SHA3-256/d' -e '/SHA3-384/d' -e '/SHA3-512/d' -e '/SHAKE128/d' -e '/SHAKE256/d' -e '/KECCAK/d' | wc -l",
-              sub { m/^Invalid Hash: 0$/ };
-            validate_script_output
-"echo -n 'Invalid Pubkey: '; openssl list -public-key-algorithms | grep '^Name' | sed -e '/RSA/d' -e '/rsa/d' -e '/DSA/d' -e '/dsa/d' -e '/EC/d' -e '/DH/d' -e '/HMAC/d' -e '/CMAC/d' | wc -l",
-              sub { m/^Invalid Pubkey: 0$/ };
+    foreach my $line (@openssl_output) {
+        chomp $line;
+        next unless ($is_openssl3 && $line =~ /$pattern->{fips}/) || ($is_openssl3 == 0 && $line =~ /$pattern->{name}/);
+
+        my $invalid_algo_found = 1;
+        foreach my $valid_algo (@valid_algos) {
+            $invalid_algo_found = 0 if $line =~ /$valid_algo/;
         }
+        die "Error: Invalid algorithm found - $line\n" if ($invalid_algo_found == 1);
```

Comment:
instead of the `for..loop` I was going to suggest something like

```perl
my @matches =  grep { $line =~ $_ } @valid_algos;
die "Error: no valid algo found" unless scalar @matches
```  

but I guess it's personal taste :)

Reply:
I'd prefer to keep the foreach with the last. It's a bit more efficient than grep (even though our dataset is very small, so it shouldn't matter much)

Reply:
speaking of efficiency: as soon as you find a matching item, you can break the loop with `last;`

Reply:
yep, I've just pushed the udpated code and doing vrs right now :)